File: creditflow/app/webhooks.py
import hmac
import hashlib
import json
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, status, Header, Request
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy import desc
from pydantic import BaseModel
from datetime import datetime

from app.database import get_db
from app.models import Tenant, WebhookEvent
from app.auth import get_current_tenant

router = APIRouter(prefix="/webhook", tags=["webhooks"])

class NEFTPayload(BaseModel):
    transaction_id: str
    sender_account: str
    receiver_account: str
    amount: float
    bank_ref: str
    timestamp: datetime

class RTGSPayload(BaseModel):
    transaction_id: str
    sender_ifsc: str
    receiver_ifsc: str
    amount: float
    purpose_code: str
    utr_number: str
    timestamp: datetime

class UPIPayload(BaseModel):
    transaction_id: str
    vpa_sender: str
    vpa_receiver: str
    amount: float
    upi_ref_id: str
    status: str
    timestamp: datetime

async def verify_signature(request: Request, tenant: Tenant):
    signature = request.headers.get("X-Webhook-Signature")
    if not signature:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Missing signature")
        
    body = await request.body()
    expected_signature = hmac.new(
        tenant.webhook_secret.encode(),
        body,
        hashlib.sha256
    ).hexdigest()
    
    if not hmac.compare_digest(expected_signature, signature):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid signature"
        )

@router.post("/neft")
async def handle_neft(
    request: Request,
    payload: NEFTPayload,
    tenant: Tenant = Depends(get_current_tenant),
    db: AsyncSession = Depends(get_db)
):
    await verify_signature(request, tenant)
    
    event = WebhookEvent(
        tenant_id=tenant.id,
        event_type="NEFT",
        payload=json.loads(payload.json()),
        status="received"
    )
    
    db.add(event)
    await db.commit()
    await db.refresh(event)
    
    return {"event_id": event.id, "status": "received"}

@router.post("/rtgs")
async def handle_rtgs(
    request: Request,
    payload: RTGSPayload,
    tenant: Tenant = Depends(get_current_tenant),
    db: AsyncSession = Depends(get_db)
):
    await verify_signature(request, tenant)
    
    event = WebhookEvent(
        tenant_id=tenant.id,
        event_type="RTGS",
        payload=json.loads(payload.json()),
        status="received"
    )
    
    db.add(event)
    await db.commit()
    await db.refresh(event)
    
    return {"event_id": event.id, "status": "received"}

@router.post("/upi")
async def handle_upi(
    request: Request,
    payload: UPIPayload,
    tenant: Tenant = Depends(get_current_tenant),
    db: AsyncSession = Depends(get_db)
):
    await verify_signature(request, tenant)
    
    # Check idempotency
    stmt = select(WebhookEvent).where(
        WebhookEvent.tenant_id == tenant.id,
        WebhookEvent.event_type == "UPI"
    )
    result = await db.execute(stmt)
    events = result.scalars().all()
    
    for e in events:
        if e.payload.get("upi_ref_id") == payload.upi_ref_id:
            return {"event_id": e.id, "status": "received", "message": "Idempotent response"}

    event = WebhookEvent(
        tenant_id=tenant.id,
        event_type="UPI",
        payload=json.loads(payload.json()),
        status="received"
    )
    
    db.add(event)
    await db.commit()
    await db.refresh(event)
    
    return {"event_id": event.id, "status": "received"}

@router.get("/events")
async def list_events(
    event_type: Optional[str] = None,
    status: Optional[str] = None,
    limit: int = 10,
    offset: int = 0,
    tenant: Tenant = Depends(get_current_tenant),
    db: AsyncSession = Depends(get_db)
):
    stmt = select(WebhookEvent).where(WebhookEvent.tenant_id == tenant.id)
    if event_type:
        stmt = stmt.where(WebhookEvent.event_type == event_type)
    if status:
        stmt = stmt.where(WebhookEvent.status == status)
        
    stmt = stmt.order_by(desc(WebhookEvent.created_at)).limit(limit).offset(offset)
    
    result = await db.execute(stmt)
    events = result.scalars().all()
    
    return {"data": events, "limit": limit, "offset": offset}

@router.get("/events/{event_id}")
async def get_event(
    event_id: str,
    tenant: Tenant = Depends(get_current_tenant),
    db: AsyncSession = Depends(get_db)
):
    result = await db.execute(
        select(WebhookEvent).where(
            WebhookEvent.id == event_id,
            WebhookEvent.tenant_id == tenant.id
        )
    )
    event = result.scalars().first()
    
    if not event:
        raise HTTPException(status_code=404, detail="Event not found")
        
    return event

File: node-app/src/webhooks.js
const express = require('express');
const crypto = require('crypto');
const { z } = require('zod');
const { WebhookEvent } = require('./models');
const { getCurrentTenant } = require('./auth');

const router = express.Router();

const NEFTPayload = z.object({
  transaction_id: z.string(),
  sender_account: z.string(),
  receiver_account: z.string(),
  amount: z.number(),
  bank_ref: z.string(),
  timestamp: z.string().datetime(),
});

const RTGSPayload = z.object({
  transaction_id: z.string(),
  sender_ifsc: z.string(),
  receiver_ifsc: z.string(),
  amount: z.number(),
  purpose_code: z.string(),
  utr_number: z.string(),
  timestamp: z.string().datetime(),
});

const UPIPayload = z.object({
  transaction_id: z.string(),
  vpa_sender: z.string(),
  vpa_receiver: z.string(),
  amount: z.number(),
  upi_ref_id: z.string(),
  status: z.string(),
  timestamp: z.string().datetime(),
});

const verifySignature = (req, res, next) => {
  const signature = req.headers['x-webhook-signature'];
  if (!signature) {
    return res.status(401).json({ detail: 'Missing signature' });
  }

  const rawBody = req.rawBody; // Assumes we set this in index.js
  const tenant = req.tenant;

  const expectedSignature = crypto
    .createHmac('sha256', tenant.webhook_secret)
    .update(rawBody)
    .digest('hex');

  if (!crypto.timingSafeEqual(Buffer.from(expectedSignature), Buffer.from(signature))) {
    return res.status(401).json({ detail: 'Invalid signature' });
  }

  next();
};

const validateBody = (schema) => (req, res, next) => {
  try {
    req.body = schema.parse(req.body);
    next();
  } catch (error) {
    return res.status(422).json({ detail: error.errors });
  }
};

router.post('/neft', getCurrentTenant(['TENANT']), verifySignature, validateBody(NEFTPayload), async (req, res) => {
  const tenant = req.tenant;
  const payload = req.body;

  const event = await WebhookEvent.create({
    tenant_id: tenant.id,
    event_type: 'NEFT',
    payload: payload,
    status: 'received'
  });

  res.json({ event_id: event.id, status: 'received' });
});

router.post('/rtgs', getCurrentTenant(['TENANT']), verifySignature, validateBody(RTGSPayload), async (req, res) => {
  const tenant = req.tenant;
  const payload = req.body;

  const event = await WebhookEvent.create({
    tenant_id: tenant.id,
    event_type: 'RTGS',
    payload: payload,
    status: 'received'
  });

  res.json({ event_id: event.id, status: 'received' });
});

router.post('/upi', getCurrentTenant(['TENANT']), verifySignature, validateBody(UPIPayload), async (req, res) => {
  const tenant = req.tenant;
  const payload = req.body;

  const existingEvent = await WebhookEvent.findOne({
    where: {
      tenant_id: tenant.id,
      event_type: 'UPI'
    }
  });

  // Since we only query by tenant and event_type, this checks if any upi_ref_id matches
  // A better approach (matching the python one) is to check if upi_ref_id matches
  const events = await WebhookEvent.findAll({
    where: {
      tenant_id: tenant.id,
      event_type: 'UPI'
    }
  });

  for (const e of events) {
    if (e.payload && e.payload.upi_ref_id === payload.upi_ref_id) {
      return res.json({ event_id: e.id, status: 'received', message: 'Idempotent response' });
    }
  }

  const event = await WebhookEvent.create({
    tenant_id: tenant.id,
    event_type: 'UPI',
    payload: payload,
    status: 'received'
  });

  res.json({ event_id: event.id, status: 'received' });
});

router.get('/events', getCurrentTenant(['TENANT']), async (req, res) => {
  const tenant = req.tenant;
  const { event_type, status, limit = 10, offset = 0 } = req.query;

  const where = { tenant_id: tenant.id };
  if (event_type) where.event_type = event_type;
  if (status) where.status = status;

  const events = await WebhookEvent.findAll({
    where,
    order: [['created_at', 'DESC']],
    limit: parseInt(limit, 10),
    offset: parseInt(offset, 10)
  });

  res.json({ data: events, limit: parseInt(limit, 10), offset: parseInt(offset, 10) });
});

router.get('/events/:event_id', getCurrentTenant(['TENANT']), async (req, res) => {
  const tenant = req.tenant;
  const event = await WebhookEvent.findOne({
    where: {
      id: req.params.event_id,
      tenant_id: tenant.id
    }
  });

  if (!event) {
    return res.status(404).json({ detail: 'Event not found' });
  }

  res.json(event);
});

module.exports = {
  router
};
